### PR TITLE
Add relative path support for include files, images handlers and testrender scenes

### PR DESCRIPTION
**Pull request by bernardkwok**
- Allow for image handler to have search paths 
- Allow for XML scene template to accept parameter overrides 
- During shader validation 
-- When reading input document, also include search path for exec + current doc path.
-- The image handler for GLSL has search paths added. For now it's the exec path + the path of the materialx file. For app integration any paths can be added.
-- The Shader is parsed to find filepaths and if relative will be converted to absolute paths and added as parameter overrides during XML scene instantiation from template.

### CODE REVIEW

File: source/MaterialXRender/Handlers/ImageHandler.cpp [hunk 1]
```diff
@@ -49,16 +51,18 @@ bool ImageHandler::saveImage(const std::string& fileName,
     return false;
 }
 
-bool ImageHandler::acquireImage(const std::string& fileName, ImageDesc &imageDesc, bool generateMipMaps, const std::array<float, 4>* /*fallbackColor*/)
+bool ImageHandler::acquireImage(const FilePath& fileName, ImageDesc &imageDesc, bool generateMipMaps, const std::array<float, 4>* /*fallbackColor*/)
 {
+    FilePath filePath = findFile(fileName);
```
Comment (bernardkwok) [inline source/MaterialXRender/Handlers/ImageHandler.cpp]: Do path search.

File: source/MaterialXRender/Handlers/ImageHandler.h [hunk 2]
```diff
@@ -207,6 +221,9 @@ class ImageHandler
     ImageLoaderMap _imageLoaders;
     /// Image description cache
     ImageDescCache _imageCache;
+
+    /// Filename search path
+    FileSearchPath _searchPath;
```
Comment (bernardkwok) [inline source/MaterialXRender/Handlers/ImageHandler.h]: Each handler instance can have a search path(s).

File: source/MaterialXRender/ShaderValidators/Osl/OslValidator.cpp [hunk 3]
```diff
@@ -113,6 +114,12 @@ void OslValidator::renderOSL(const string& outputPath, const string& shaderName,
     replacementMap[OUTPUT_SHADER_TYPE_STRING] = outputShader;
     replacementMap[OUTPUT_SHADER_INPUT_STRING] = OUTPUT_SHADER_INPUT_VALUE_STRING;
     replacementMap[INPUT_SHADER_TYPE_STRING] = shaderName;
+    string overrideString;
+    for (auto param : _oslShaderParameterOverrides)
```
Comment (bernardkwok) [inline source/MaterialXRender/ShaderValidators/Osl/OslValidator.cpp]: Bind any parameter overrides. Assumes the string(s) have the correct syntax.

File: source/MaterialXTest/ShaderValid.cpp [hunk 4]
```diff
@@ -868,6 +872,40 @@ static void runOSLValidation(const std::string& shaderName, mx::TypedElementPtr
...
+                        if (uniform->type != MaterialX::Type::FILENAME)
+                        {
+                            continue;
+                        }
+                        if (uniform->value)
+                        {
+                            const std::string& uniformName = uniform->name;
+                            mx::FilePath filename;
+                            mx::FilePath origFilename(uniform->value->getValueString());
+                            if (!origFilename.isAbsolute())
```
Comment (bernardkwok) [inline source/MaterialXTest/ShaderValid.cpp]: For each input filename try and find the absolute path name if is relative. If found then set parameter overrides for the shader.

File: source/MaterialXTest/ShaderValid.cpp [hunk 5]
```diff
@@ -1437,7 +1475,7 @@ TEST_CASE("MaterialX documents", "[shadervalid]")
             const std::string filename = filePath;
 
             mx::DocumentPtr doc = mx::createDocument();
-            readFromXmlFile(doc, filename);
+            mx::readFromXmlFile(doc, filename, dir);
```
Comment (bernardkwok) [inline source/MaterialXTest/ShaderValid.cpp]: Include document directory + exec (above documents/...) as search path. Will handle xincludes.

File: documents/TestSuite/Utilities/closure_color_scene.xml [hunk 6]
```diff
@@ -26,12 +26,14 @@
    <Quad corner="-100,100,0" edge_x="0,0,200" edge_y="200,0,0" is_light="yes"/>
 
    <!-- Shader graph for routing to output layer:
+   input_shader_parameter_overrides : are parameter overrides for the input shader
```
Comment (feldstj) [inline documents/TestSuite/Utilities/closure_color_scene.xml]: Don't you need the %% around input_shader_parameter_overrides?

File: documents/TestSuite/Utilities/constant_color_scene.xml [hunk 7]
```diff
@@ -13,12 +13,14 @@
    <Background resolution="1024" />
    -->
    <!-- Shader graph for routing to output layer:
+   input_shader_parameter_overrides : are parameter overrides for the input shader
```
Comment (feldstj) [inline documents/TestSuite/Utilities/constant_color_scene.xml]: smae here, the %%

**Pull request merged**